src/compiler/checker.ts

```diff
@@ -29007,18 +29007,45 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
-                isTupleType(t) ?
-                    getElementTypeOfSliceOfTupleType(t, 0, /*endSkipCount*/ 0, /*writing*/ false, /*noReductions*/ true) :
-                    getIteratedTypeOrElementType(IterationUse.Element, t, undefinedType, /*errorNode*/ undefined, /*checkAssignability*/ false),
-                /*noReductions*/ true));
+    function getSpreadIndices(elements: readonly Node[]) {
+        let first, last;
+        for (let i = 0; i < elements.length; i++) {
+            if (isSpreadElement(elements[i])) {
+                first ??= i;
+                last = i;
+            }
+        }
+        return { first, last };
+    }
+
+    function getContextualTypeForElementExpression(type: Type | undefined, index: number, length?: number, firstSpreadIndex?: number, lastSpreadIndex?: number): Type | undefined {
+        return type && mapType(type, t => {
+            if (isTupleType(t)) {
+                // If index is before any spread element and within the fixed part of the contextual tuple type, return
+                // the type of the contextual tuple element.
+                if ((firstSpreadIndex === undefined || index < firstSpreadIndex) && index < t.target.fixedLength) {
+                    return getTypeArguments(t)[index];
+                }
+                // When the length is known and the index is after all spread elements we compute the offset from the element
+                // to the end and the number of ending fixed elements in the contextual tuple type.
+                const offset = length !== undefined && (lastSpreadIndex === undefined || index > lastSpreadIndex) ? length - index : 0;
+                const fixedEndLength = offset > 0 && t.target.hasRestElement ? getEndElementCount(t.target, ElementFlags.Fixed) : 0;
+                // If the offset is within the ending fixed part of the contextual tuple type, return the type of the contextual
+                // tuple element.
+                if (offset > 0 && offset <= fixedEndLength) {
+                    return getTypeArguments(t)[getTypeReferenceArity(t) - offset];
+                }
+                // Return a union of the possible contextual element types with no subtype reduction.
+                return getElementTypeOfSliceOfTupleType(t,
+                    firstSpreadIndex === undefined ? t.target.fixedLength : Math.min(t.target.fixedLength, firstSpreadIndex),
+                    length === undefined || lastSpreadIndex === undefined ? fixedEndLength : Math.min(fixedEndLength, length - lastSpreadIndex),
+                    /*writing*/ false, /*noReductions*/ true);
+            }
+            // If element index is known and a contextual property with that name exists, return it. Otherwise return the
+            // iterated or element type of the contextual type.
+            return (!firstSpreadIndex || index < firstSpreadIndex) && getTypeOfPropertyOfContextualType(type, "" + index as __String) ||

```

Comment:
Maybe it should be treated as a separate issue... but I feel that perhaps this feature here might directly complicate things for that issue so it's at least worth mentioning now.

Take a look at this [TS playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMAeAFXhAA9DVgj40BrVHAd1QG0BdAGnjkATFVoh6jFm068AfAAoAUPHhQAXPC7KV8AHT6A3tp2aA0vCz5mIAJ45EgnhqMmTABxg5gyMCA3yiEBBgDRIYSwBzAEp4AF5ZQS5THgBuYx0AXzSTDO1+bQAjDS0TfV0XVySLK1t7QSEneArXeDA8ImQAWz94eVgIjXIqgDJ4azsHch4Y+Pqk1PSVLPTclR5FKI0ANxwsYDTFQhIlFRKdZp0PLx8e+RmEgCIHvhWXkwuVK+9ff3v4ABYhG9MsD8qd0h9Wu0urd+n9ISo2qgSPA4L4sFtgqEMOFUBE4qoYBFsi1USAMMgYPg0SAMcESTlgUsmU1FlDkTD-HC4gkEeyUTS6SF4KgugUQDACf0Ga44BSqWT0Zj9myMiy1XkNmkgA). The problem there is that [`getTypeOfPropertyOfContextualType`](https://github.dev/microsoft/TypeScript/blob/b2f0543bcaadaf5ce3efd4cb33c802dfe3c4e71c/src/compiler/checker.ts#L28947) gets the [`getElementTypeOfSliceOfTupleType`](https://github.dev/microsoft/TypeScript/blob/b2f0543bcaadaf5ce3efd4cb33c802dfe3c4e71c/src/compiler/checker.ts#L28963-L28966), so we end up with a union of all element types instead of types of individual tuple members.

I prepared a quick fix for this issue in this PR: https://github.com/microsoft/TypeScript/pull/53042 but it doesn't handle the case that is being fixed here (ending tuple elements).

The case outlined there is currently a little bit silly (`T[K & keyof T]` within `[K in keyof T2]: { ... }`) but I partially encountered this need when working on inference for intersected mapped types ([PR here](https://github.com/microsoft/TypeScript/pull/52062)). This case would be better if rewritten like this:
```diff
diff --git a/tests/cases/compiler/contextuallyTypedElementsOfGenericZippingTuples.ts b/tests/cases/compiler/contextuallyTypedElementsOfGenericZippingTuples.ts
index a011321a26..c27c91930f 100644
--- a/tests/cases/compiler/contextuallyTypedElementsOfGenericZippingTuples.ts
+++ b/tests/cases/compiler/contextuallyTypedElementsOfGenericZippingTuples.ts
@@ -11,8 +11,8 @@ declare function test<T extends unknown[], T2 extends unknown[]>(
   ],
   b: [
     ...{
-      [K in keyof T2]: {
-        consume: (arg: T[K & keyof T]) => T2[K];
+      [K in keyof T2 as K & keyof T]: {
+        consume: (arg: T[K]) => T2[K];
       };
     }
   ]
```

There are 2 problems with this code today though:
1. such a reverse mapped type isn't inferred at all (a PR fixing this problem can be found [here](https://github.com/microsoft/TypeScript/pull/52972))
2. this isn't recognized as a tuple type so TS reports `A rest element type must be an array type.ts(2574)`. That probably could be improved as well.

Reply:
@ahejlsberg Any concerns given the above?

Reply:
Yes, I just meant specifically the "this feature here might directly complicate things for that issue so it's at least worth mentioning now" bit.